### ISSUE TITLE
Does the current version (codes instead of checkpoints) use knowledge insulation during fine-tuning when using train.py?

### ISSUE THREAD

**Issue by hongmingli1995** [Physical-Intelligence/openpi#649]
Thanks for sharing! I'm not sure if I’m understanding things correctly, but it seems that when I fine-tune using script/train.py, both _paligemma_variant_ and _action_expert_variant_ are being trained, and knowledge insulation isn't being applied.

You mentioned it “only supports the flow matching head” — does that mean I should set paligemma_variant = None?

**Comment by kpertsch** [comment]
Your understanding is correct: openpi doesn't currently implement knowledge insulation for finetuning (or combined FAST + action expert flow matching losses for that matter). openpi finetuning currently only supports finetuning with the action expert flow matching loss and gradients get propagated through *both* the action expert and the VLM backbone.
Hope that answers your question!

**Comment by zwbx** [comment]
Hi, I have a follow-up question that is it reasonable to do  `knowledge insulation` in finetune state? That being said the VLM part is frozen and only tune the action expert.

Also would like make sure if is it true: during pretraining state, `knowledge insulation` is actually enabled according to your paper.

**Comment by kpertsch** [comment]
Indeed, we use KI during pre-training for the pi05 checkpoints.
During finetuning, it likely will work a lot worse if you freeze the full VLM backbone and only finetune the action expert (you can try though). You *can* use KI during finetuning, but it will likely work much better if you still update the VLM backbone using FAST token targets (ie discrete action token prediction).
That being said, openpi doesn't implement this atm, and we haven't seen conclusive evidence that it actually helps during finetuning.

**Comment by taogoddd** [comment]
Thanks for your reply. I am a bit confused about KI here. I thought KI means combined FAST + action expert flow matching losses + stop gradient in between, and you guys mentioned that this kind of KI helps during both pre-training and finetuning. But you said in the end that no conclusive evidence that it actually helps during finetuning. So:
1. Is it advised to implement KI during fine-tuning? (It seems the pi05-droid ckpt uses this)
2. Do you have a plan to release related codes for KI?

Thanks!

> Indeed, we use KI during pre-training for the pi05 checkpoints. During finetuning, it likely will work a lot worse if you freeze the full VLM backbone and only finetune the action expert (you can try though). You _can_ use KI during finetuning, but it will likely work much better if you still update the VLM backbone using FAST token targets (ie discrete action token prediction). That being said, openpi doesn't implement this atm, and we haven't seen conclusive evidence that it actually helps during finetuning.
